plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,241 @@
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return createQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                restoreInflightValues(queryGroup.getResourceLimits());
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                restoreInflightValues(queryGroup.getResourceLimits());
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * Get the allocation value for resourceName for the QueryGroup
+     * @param resourceName - the resourceName we want to get the usage for
+     * @param resourceLimits  - the resource limit from which to get the allocation value for resourceName
+     */
+    private double getResourceLimitValue(String resourceName, final Map<ResourceType, Object> resourceLimits) {
+        for (ResourceType resourceType : resourceLimits.keySet()) {
+            if (resourceType.getName().equals(resourceName)) {
+                return (double) resourceLimits.get(resourceType);
+            }
+        }
+        return 0.0;
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
+        final Metadata metadata = currentClusterState.metadata();
+        String groupName = queryGroup.getName();
+        final Map<String, QueryGroup> previousGroups = metadata.queryGroups();
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        String resourceNameWithThresholdExceeded = "";
+        for (ResourceType resourceType : queryGroup.getResourceLimits().keySet()) {
+            String resourceName = resourceType.getName();
+            double existingUsage = calculateExistingUsage(resourceName, previousGroups, groupName);
+            double newGroupUsage = getResourceLimitValue(resourceName, queryGroup.getResourceLimits());
+            inflightResourceLimitValues.computeIfAbsent(resourceName, k -> new DoubleAdder()).add(newGroupUsage);
+            double totalUsage = existingUsage + inflightResourceLimitValues.get(resourceName).doubleValue();
+            if (totalUsage > 1) {
```

Comment:
Do we always want this as 1 ? Make it configurable?
In addition to this, might be a better idea to move all these validation logic into respective separate class/component where they optionally return an exception in case validation was unsuccessful (just an idea). As currently this method looks bloated.

Reply:
Yeah this should be 1 all the time. This is percentage so the max allowed is 1

Reply:
Not sure if this would be 1 all the way through. We can later decide to keep it at 0.9 maybe to have some buffer in which case it makes sense to have this a static setting. @kaushalmahi12 @jainankitk What do you think?

Reply:
Can you elaborate little more on this ?

Reply:
We have a node level rejection and cancelation threshold, so the values here will be related to that. For example, if the rejection threshold is 90%, then 1 here will be 1*90%=90%. It won't bypass 90%